### PR TITLE
Uses blink-alloc instead of bumpalo

### PR DESCRIPTION
**Pull request by zakarumych**
The usage is roughly the same.

One caveat applies:
String is not generalized over allocator in std, so `allocator-api2` lacks one too.

There seems to be room for perf improvement.
On my machine this change gives 2.5 - 3% speed increase

### CODE REVIEW

**Comment by Boshen** [comment]
The test failed with 

```
thread 'no_bloat_enum_sizes' panicked at 'assertion failed: `(left == right)`
  left: `24`,
 right: `16`', crates/oxc_ast/src/lib.rs:58:5
```

with

https://github.com/Boshen/oxc/blob/47c1e7ef5124b81d6df0cd8182b56a94e297a04b/crates/oxc_ast/src/lib.rs#L58

Which is fine I think, we can mark it as 24.

**Comment by Boshen** [comment]
Here are the benchmarks from the CI jobs, they aren't posted by bot due to permission issues.

### Parser Benchmark Results - macos-latest
```
group                    main                                   pr
-----                    ----                                   --
parser/babylon.max.js    1.05   192.8±50.76ms    53.6 MB/sec    1.00   182.8±19.32ms    56.5 MB/sec
parser/d3.js             1.12    25.2±32.18ms    21.7 MB/sec    1.00     22.5±4.48ms    24.3 MB/sec
parser/lodash.js         1.00      5.6±0.44ms    91.6 MB/sec    1.11      6.2±1.41ms    82.6 MB/sec
parser/pdf.js            1.00     11.5±0.69ms    34.9 MB/sec    1.14     13.1±2.60ms    30.6 MB/sec
parser/typescript.js     1.00   182.9±28.20ms    52.6 MB/sec    1.01   184.1±35.35ms    52.2 MB/sec
```

### Parser Benchmark Results - windows-latest
```
group                    main                                   pr
-----                    ----                                   --
parser/babylon.max.js    1.00    138.7±3.34ms    74.4 MB/sec    1.08    150.4±2.21ms    68.7 MB/sec
parser/d3.js             1.00     16.7±0.38ms    32.8 MB/sec    1.07     17.9±0.21ms    30.6 MB/sec
parser/lodash.js         1.00      4.7±0.19ms   109.6 MB/sec    1.09      5.1±0.10ms   100.6 MB/sec
parser/pdf.js            1.00      9.4±0.21ms    42.7 MB/sec    1.09     10.3±0.23ms    39.0 MB/sec
parser/typescript.js     1.00    137.1±2.33ms    70.2 MB/sec    1.10    150.2±2.09ms    64.0 MB/sec
```

### Parser Benchmark Results - ubuntu-latest
```
group                    main                                   pr
-----                    ----                                   --
parser/babylon.max.js    1.00    126.2±3.95ms    81.8 MB/sec    1.15    145.3±4.75ms    71.1 MB/sec
parser/d3.js             1.00     15.1±0.58ms    36.2 MB/sec    1.20     18.1±0.50ms    30.3 MB/sec
parser/lodash.js         1.00      4.4±0.16ms   117.8 MB/sec    1.15      5.0±0.26ms   102.9 MB/sec
parser/pdf.js            1.00      8.9±0.26ms    45.3 MB/sec    1.13     10.0±0.35ms    40.2 MB/sec
parser/typescript.js     1.00    124.0±3.70ms    77.6 MB/sec    1.18    145.9±4.53ms    66.0 MB/sec
```

**Comment by Boshen** [comment]
The performance gains are really significant, and I love it! Your allocator is amazing!

> There seems to be room for perf improvement.
> On my machine this change gives 2.5 - 3% speed increase

I'm free to either merge this or wait for your approval.

**Comment by Boshen** [comment]
@YoniFeng Would also like to review this? I'm also a Rust noob here, we can exchange notes ;-)

File: crates/oxc_allocator/src/arena.rs [hunk 1]
```diff
@@ -8,155 +8,9 @@ use std::{
...
-    #[must_use]
-    pub fn unbox(self) -> T {
-        // This pointer read is safe because the reference `self.0` is
-        // guaranteed to be unique--not just now, but we're guaranteed it's not
-        // borrowed from some other reference. This in turn is because we never
-        // construct an alloc::Box with a borrowed reference, only with a fresh
-        // one just allocated from a Bump.
-        unsafe { ptr::read(self.0 as *mut T) }
-    }
-}
```
Comment (Boshen) [inline crates/oxc_allocator/src/arena.rs:30]: I kind of liked the `unbox` API, its from jsparagus

https://github.com/mozilla-spidermonkey/jsparagus/blob/939577de8fb80c94c4a1595a2be69802327be680/crates/ast/src/arena.rs#L13-L22

File: crates/oxc_cli/src/lib.rs [hunk 2]
```diff
@@ -126,7 +126,7 @@ impl Cli {
         let parser_source_text = source_text.clone();
         let ret = Parser::new(&allocator, &parser_source_text, source_type).parse();
         let result = if ret.errors.is_empty() {
-            let program = allocator.alloc(ret.program);
+            let program = allocator.emplace_no_drop().value(ret.program);
```
Comment (Boshen) [inline crates/oxc_cli/src/lib.rs:129]: These `emplace-*` APIs are pretty new to me.

**Comment by Boshen** [comment]
I'll clean this up and merge when @zakarumych and @YoniFeng gives a 👍

**Comment by Boshen** [comment]
Wait ... it seems to be slower from the benchmarks, I didn't fully wake up yet and saw it the other way around.

File: crates/oxc_allocator/src/lib.rs [hunk 3]
```diff
@@ -1,30 +1,80 @@
...
+    blink: Blink,
 }
 
 // SAFETY: Make Bump Sync and Send, it's our responsibility to never
 // simultaneously mutate across threads.
 unsafe impl Send for Allocator {}
 unsafe impl Sync for Allocator {}
 
+pub type Box<'a, T> = allocator_api2::boxed::Box<T, &'a BlinkAlloc>;
+pub type Vec<'a, T> = allocator_api2::vec::Vec<T, &'a BlinkAlloc>;
```
Comment (YoniFeng) [inline crates/oxc_allocator/src/lib.rs:17]: I think this is the main piece of the puzzle I was missing.
You can annotate the lifetime regardless of `allocator_api2::vec::Vec/Box` having a signature "set up for it".

**Comment by YoniFeng** [comment]
Will need to profile if we want to understand why its slower.


The internals complexity is beyond understanding with a code skim, one thing that stood out to me was was that the `unbox()` was just dereferencing a pointer, and now it's:
```rust
pub fn into_inner(boxed: Self) -> T {
        let ptr = boxed.0;
        let unboxed = unsafe { ptr.as_ptr().read() };
        unsafe { boxed.1.deallocate(ptr.cast(), Layout::new::<T>()) };
        unboxed
    }
```

Deallocate is trying to reclaim the memory if the deallocated object is the last one.
I guess after inlining this boils down to a simple `add` and `cmp`. While I don't think it's significant/can explain the worse performance overall, it is an example of a place where more work is done.

**Comment by zakarumych** [comment]
I was thinking of removing deallocations or making them only for large enough blocks

**Comment by Boshen** [comment]
@zakarumych Are the benchmarks wrong or is there something fishy going on?

File: crates/oxc_allocator/src/arena.rs (hunk 1 above)
Comment (zakarumych) [reply crates/oxc_allocator/src/arena.rs:30]: `Box::into_inner` is from `std`.
And `Box` should not have methods with `self` to not cause name resolution conflict with pointee.

File: crates/oxc_cli/src/lib.rs (hunk 2 above)
Comment (zakarumych) [reply crates/oxc_cli/src/lib.rs:129]: `Blink` allows dropping values placed into the allocated memory.
`emplace_*` configure how values are placed.

**Comment by zakarumych** [comment]
Turned out that duplicates of `std`s `Box` and `Vec` are problematic.
`bumpalo`'s duplicates spray `#[inline]` and `#[inline(always)]` all over the place.
`allocator-api2`'s duplicates are almost char-to-char copy with no additional inlining attributes.

**Comment by Boshen** [comment]
> Turned out that duplicates of `std`s `Box` and `Vec` are problematic. `bumpalo`'s duplicates spray `#[inline]` and `#[inline(always)]` all over the place. `allocator-api2`'s duplicates are almost char-to-char copy with no additional inlining attributes.

Is there anything I can help out with?

**Comment by Boshen** [comment]
@zakarumych Friendly ping :-) May I ask you what the current status is?

**Comment by Boshen** [comment]
Closing due to inactivity. Please reach out if you want to work on this again. And thank you for working on this, I learned a bunch from your work.

**Pull request closed without merging**